### PR TITLE
cacher: returns an error when watch list was requested and storage.RequestWatchProgress is disabled

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 1]
```diff
@@ -2246,7 +2246,7 @@ func TestGetWatchCacheResourceVersion(t *testing.T) {
 				require.NoError(t, err)
 			}
 
-			actualResourceVersion, err := cacher.getWatchCacheResourceVersion(context.TODO(), uint64(parsedResourceVersion), scenario.opts)
+			actualResourceVersion, err := cacher.getWatchCacheResourceVersion(context.TODO(), uint64(parsedResourceVersion), scenario.opts, true)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:2249]: The `TestGetWatchCacheResourceVersion` tests all combinations of `ResourceVersion`, `AllowWatchBookmarks` and `SendInitialEvents`.

Once we are okay with how `getWatchCacheResourceVersion` works we could extend this test by `isRequestWatchProgressSupported`.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 2]
```diff
@@ -1322,14 +1326,17 @@ func (c *Cacher) getBookmarkAfterResourceVersionLockedFunc(parsedResourceVersion
...
+func (c *Cacher) getWatchCacheResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions, isRequestWatchProgressSupported bool) (uint64, error) {
 	if len(opts.ResourceVersion) != 0 {
 		return parsedWatchResourceVersion, nil
 	}
 	// legacy case
 	if !utilfeature.DefaultFeatureGate.Enabled(features.WatchFromStorageWithoutResourceVersion) && opts.SendInitialEvents == nil && opts.ResourceVersion == "" {
 		return 0, nil
 	}
+	if isListWatchRequest(opts) && !isRequestWatchProgressSupported {
+		return 0, fmt.Errorf("a watch stream was requested by the client but the required storage feature %s is disabled", storage.RequestWatchProgress)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1338]: this error will be wrapped into `newErrWatcher` and  which translates to `http.StatusInternalServerError`, we could return a different error for example `errors.NewMethodNotSupported(...)`

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -1322,14 +1326,17 @@ func (c *Cacher) getBookmarkAfterResourceVersionLockedFunc(parsedResourceVersion
...
-func (c *Cacher) getWatchCacheResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (uint64, error) {
+func (c *Cacher) getWatchCacheResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions, isRequestWatchProgressSupported bool) (uint64, error) {
 	if len(opts.ResourceVersion) != 0 {
 		return parsedWatchResourceVersion, nil
 	}
 	// legacy case
 	if !utilfeature.DefaultFeatureGate.Enabled(features.WatchFromStorageWithoutResourceVersion) && opts.SendInitialEvents == nil && opts.ResourceVersion == "" {
 		return 0, nil
 	}
+	if isListWatchRequest(opts) && !isRequestWatchProgressSupported {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1337]: I'm not sure that's the right place - you can have watch-list request also with a given RV (line 1330), and we probably want to fail for such case too.

I think that this check should be moved out of this function and be called directly before `getWatchCacheResourceVersion` function is called.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1337]: ok, good point, thanks.

**Pull request merged**